numpy/polynomial/polyutils.py

```diff
@@ -684,6 +684,264 @@ def _fit(vander_f, x, y, deg, rcond=None, full=False, w=None):
     else:
         return c
 
+def _get_coeff_idx(shape):
+    """
+    Get the powers of x and y corresponding to a 2d coefficient matrix
+    in the same order as the Vandermonde matrix,
+    i.e. [[0, 0], [1, 0], ...]
+
+    Parameters
+    ----------
+    shape : 2-tuple
+        shape of the coefficient matrix, i.e. degree + 1
+
+    Returns
+    -------
+    idx : array of shape (n * m, 2)
+        pairs of indices to the coefficient matrix
+    """
+    assert len(shape) == 2, "Expected the shape of a 2d array"
+    idx = np.indices(shape)
+    idx = idx.T.swapaxes(0, 1).reshape((-1, 2))
+    return idx
+
+
+def _scale(*args):
+    """
+    Scale inputs independently to mean 0 and variance 1
+
+    Parameters
+    ----------
+    *args : array
+        any number of arrays
+
+    Returns
+    -------
+    *x : array
+        scaled arrays
+    norm : array of shape (n,)
+        scaling factors
+    offset : array of shape (n,)
+        offsets that were subtracted
+    """
+    # Normalize x and y to avoid huge numbers
+
+    n = len(args)
+    norm = [None for _ in range(n)]
+    out = [None for _ in range(n)]
+    for i, x in enumerate(args):
+        # need to convert to float in case its an integer
+        # Also make sure its a numpy array
+        norm[i] = np.max(np.abs(x))
+        if norm[i] == 0:
+            norm[i] = 1
+        out[i] =  x / norm[i]
+    return (*out, norm)
+
+
+def _unscale(*args, norm):
+    """
+    Invert the scaling on any number of arrays
+
+    Parameters
+    ----------
+    norm : array of shape (n,)
+        scaling factors
+    offset : array of shape (n,)
+        applied offsets
+
+    Returns
+    -------
+    *x : array
+        arrays with inverted scaling
+    """
+    n = len(args)
+    out = [None for _ in range(n)]
+    for i, x in enumerate(args):
+        x *= norm[i]
+        out[i] = x
+    return out
+
+
+def polyscale2d(coeff, scale_x, scale_y, copy=True):
+    """
+    Scale the coefficients of a 2d polynomial by scaling factors
+    previously applied to the coordinates
+
+    Parameters
+    ----------
+    coeff : array_like
+        polynomial coefficients
+    scale_x : float
+        scaling factor in x direction
+    scale_y : float
+        scaling factor in y direction
+    copy : bool, optional
+        Whether to copy the coefficients, by default True
+
+    Returns
+    -------
+    coeff : array_like
+        scaled coefficients
+    """
+    if copy:
+        coeff = np.copy(coeff)
+    idx = _get_coeff_idx(coeff.shape)
+    for i, j in idx:
+        coeff[i, j] /= scale_x ** i * scale_y ** j
+    return coeff
```

Comment:
I think this can be vectorized, will come back to it...

Reply:
Done